### PR TITLE
Feature/save post for later native caching

### LINKED ISSUE #7699: Save posts for later - implement caching of posts
Implement caching of posts saved for later using native WebView cache implementation or by saving their media content locally.

### PR DESCRIPTION
**Pull request by khaykov**
Fixes #7699 

For Phase 1 we are going to utilize native WebView cache. Simple WebView cache is a "best-effort" kind-of cache, as we can't guarantee that it will stick for a long time, or if we would be able to fully cache post media. 

WebView caching runs on UI thread, so I'm running it in a dedicated fragment (not attached to any container). Because orientation change destroys UI, caching will be restarted from scratch. 

Caching is quiet, and we do not notify users of progress or errors.
Caching is not performed when "Saved posts" filter is selected.

To test:
- Bookmark a post you haven't opened before (clear cache for a sanity check) with images in it.
- Turn off the internet and open the post - images should be in it.

Next steps:
For Phase 2 we have a couple of options:

- Utilize [HTML5 App Cache API](https://www.html5rocks.com/en/tutorials/appcache/beginner/). If we can implement this server-side this might be the easiest way to do cache inside the app, as WebView supports it natively. Alternatively, we can try to create and inject the manifest manually, but this is quite hacky.

- Utilize WebViewClient to intercept loading of resources and cache images manually. This is a bit more involved process, but we will have more control over cache and will be able to store it as app data (more persistent).

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostWebViewCachingFragment.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        args.putLong(ARG_BLOG_ID, blogId);
+        args.putLong(ARG_POST_ID, postId);
+        fragment.setArguments(args);
+        return fragment;
+    }
+
+    @Override public void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        setRetainInstance(true);
```
Comment (malinajirka) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostWebViewCachingFragment.java]: What are some benefits of using retained fragment in this case?

**Comment by malinajirka** [comment]
Great job @khaykov! I really like the idea of separating the pre-loading logic into another fragment!

Couple of notes - none of them is blocking, however I'd like to bring them up for consideration.
1. Pre-loading images doesn't work for all posts. I've tried to pre-load few posts with 'Automattic' tag and it didn't work. I've noticed they all had `cache-control: private` header (afaik it's a header mostly intended for proxies so I'm really not sure it's related). Weird thing is that if I open the post detail while online, the images are properly cached. 

2. >For Phase 2 we have a couple of options:
It's awesome that you've started investigating options, how to improve the caching strategy! However I'd consider keeping the same caching strategy even for the phase 2. I've tried to save 8 posts and then open another 10 posts with images and the cache of the 8 saved posts didn't get overridden. I believe it should be enough for most use cases. The implementation of persistent image cache might be quite complicated and I think we should make sure it's really needed first, before we dive into it. WDYT?

3. There is no need to rewrite already implemented code, but we might want to start writing new classes in Kotlin and even consider using ViewModel pattern for new fragments/activities. WDYT?

Thank you for this PR!!! 👍

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostWebViewCachingFragment.java (hunk 1 above)
Comment (khaykov) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostWebViewCachingFragment.java]: Oops. I used it with initial implementation and forgot to remove. Good catch!

**Comment by khaykov** [comment]
I addressed the issue with image caching in private posts, but I can't seem to make it work reliably with 3rd party media, especially when it comes to large images or gifs.

Not exactly sure what the issue might be, but when loading big images, if we put a breakpoint [here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderWebView.java#L258), and just go through it, the caching will work. I suspect it might be OOM related. 


> It's awesome that you've started investigating options, how to improve the caching strategy! However I'd consider keeping the same caching strategy even for the phase 2. I've tried to save 8 posts and then open another 10 posts with images and the cache of the 8 saved posts didn't get overridden. I believe it should be enough for most use cases. The implementation of more persistent image cache might be quite complicated and I think we should make sure it's really needed first, before we dive into it. WDYT?

I absolutely agree with you. My suggestions are more like a memo on research I did. We don't need to act on them if the current implementation is enough :)

> There is no need to rewrite already implemented code, but we might want to start writing new classes in Kotlin and even consider using ViewModel pattern for new fragments/activities. WDYT?

Tottaly forgot that we can now use Kotlin in main app 🤦‍♂️

**Comment by malinajirka** [comment]
Really great job @khaykov ! :)!!!

>I absolutely agree with you. My suggestions are more like a memo on research I did. We don't need to act on them if the current implementation is enough :)
 
👍 

>Tottaly forgot that we can now use Kotlin in main app 🤦‍♂️

No worries at all:D.


I tried to reproduce the issue with big images, but had no luck. It works for me. I've added custom 'photography' tag and bookmarked several articles with huge photos and everything worked as expected. I even tried to change the network type (on an emulator) to some slower connection, but everything worked. I believe this PR is great improvement already and if we decide to tackle the big images issue, we can create a new github issue.

**Pull request merged**